docs(salesforce-jwt): add provider and customer guides for salesforceJWT (#562)
amp-labs/docs: src/customer-guides/salesforce-jwt.mdx

---

> title: "Salesforce (JWT)"

Reviewer: ```suggestion
title: "Salesforce (JWT authentication)"
```

---

> This guide walks you through connecting your Salesforce org to an integration that uses the OAuth 2.0 JWT Bearer Flow. This is a headless, server-to-server authentication method — there is no interactive login, and the integration runs on behalf of a designated user using credentials you configure up front.

Reviewer: can you link to Salesforce docs

> If your integration uses the standard Salesforce OAuth flow instead, please refer to the [Salesforce customer guide](/customer-guides/salesforce).

Reviewer: ```suggestion
If your integration uses the standard Salesforce OAuth flow instead, please refer to the [Salesforce customer guide](/customer-guides/salesforce). If you are unsure which guide is applicable to you, please ask the integration developer.
```

## Before installing a Salesforce (JWT) integration

You will need **Administrator access** to your Salesforce org.

### 1. Generate a certificate and RSA private key

> The JWT Bearer flow requires an RSA key pair. You'll upload the public certificate to your Salesforce Connected App and share the private key (base64-encoded) with the integration.

Reviewer: ```suggestion
The JWT Bearer flow requires an RSA key pair. You'll upload the public certificate to your Salesforce External Client App and share the private key (base64-encoded) with the integration.
```

  -keyout private-key.pem -out cert.pem \
  -subj "/CN=salesforce-jwt-integration"
```

This produces two files:
- `private-key.pem` — Your RSA private key. You will base64-encode this and share it when installing the integration.
- `cert.pem` — Your public certificate. You will upload this to Salesforce in a later step.

<Note>Salesforce requires an RSA key. ECDSA (`ecparam`) keys are not accepted for the JWT Bearer flow. Use a key length of at least 2048 bits.</Note>

You will then need to base64-encode your private key before sharing it during the installation process. Use the command that matches your environment:

**macOS (Terminal)**

```bash
base64 -i private-key.pem
```

**Linux, Git Bash, or WSL**

```bash
base64 -w 0 private-key.pem
```

**Windows (PowerShell)**

From the directory that contains `private-key.pem`:

```powershell
[Convert]::ToBase64String([IO.File]::ReadAllBytes((Resolve-Path 'private-key.pem')))
```

Copy the encoded string (a long single line, e.g. starting with `LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tL...`). It will be required during the installation process.

### 2. Create the app in Salesforce

> The app identifies the integration to Salesforce and holds the certificate used to verify JWT signatures. Since Spring '26, Salesforce no longer allows creating new **Connected Apps** — use an **External Client App** instead. If you already have a Connected App from before the change, it continues to work and you can use it as-is.

Reviewer: let's just only document External Client Apps, I doubt any consumers are going to be recycling old Connected Apps to connect to a new integration, so we don't have to consider Connected Apps.

<Tabs>
  <Tab title="External Client App">

Use this path if you are creating a new app in Spring '26 or later.

1. Log in to *Salesforce* as an Administrator.
2. Go to **Setup**.
3. In the *Quick Find* box, enter **External** and select **External Client App Manager**.
4. Click **New External Client App**.
5. Fill in the basic information:
   - **External Client App Name**: e.g. `Ampersand Integration`
   - **API Name**: auto-filled from the name
   - **Contact Email**: your administrator email
   - **Distribution State**: **Local**
6. Under **API (Enable OAuth Settings)**:
   - Check **Enable OAuth**.
   - For **Callback URL**, enter `https://api.withampersand.com/callbacks/v1/oauth`. The JWT Bearer flow does not use the callback, but Salesforce requires a value.
   - Under **Selected OAuth Scopes**, add:
     - **Manage user data via APIs (`api`)**
     - **Perform requests at any time (`refresh_token`, `offline_access`)** — this is required for Salesforce to accept JWT Bearer exchanges even though no refresh token is issued.

>      - Alternatively, add **Full access** (`full`) if you want to grant all permissions to the integration.

Reviewer: remove this sentence, people that want this flow are security conscious and would never grant `full`

4. Under **Plugin Policies**, set **Permitted Users** to **Admin approved users are pre-authorized**.
5. Under **App Authorization**:
   - Set **Refresh Token Policy** to **Refresh token is valid until revoked**.
   - Set **IP Relaxation** to **Relax IP restrictions**.
6. Click **Save**.

<Note>Setting **Permitted Users** to **Admin approved users are pre-authorized** is required for the JWT Bearer flow. The flow is headless — the integration user never completes an interactive OAuth consent — so the admin must authorize them ahead of time by assigning a permission set to the app. Without this, Salesforce rejects the JWT assertion with `user hasn't approved this consumer`.</Note>

<Note>Setting **Refresh Token Policy** to **Refresh token is valid until revoked** is also required, even though the JWT Bearer flow never actually issues a refresh token. Salesforce internally checks for a non-expired approval record for the user before issuing an access token, and any other setting causes that approval to age out — which silently breaks the integration later. The other options (`Immediately expire`, `Expire after specific time`, etc.) will cause `invalid_grant` errors.</Note>

  </Tab>

>   <Tab title="Connected App (legacy)">

Reviewer: can remove this tab

1. Go back to the Connected App detail page.
2. Click **Manage** at the top.
3. Click **Edit Policies**.
4. Under **OAuth Policies**:
   - Set **Permitted Users** to **Admin approved users are pre-authorized**.
   - Set **IP Relaxation** to **Relax IP restrictions**.
   - Set **Refresh Token Policy** to **Refresh token is valid until revoked**.
5. Click **Save**.

<Note>Setting **Permitted Users** to **Admin approved users are pre-authorized** is required for the JWT Bearer flow. The flow is headless — the integration user never completes an interactive OAuth consent — so the admin must authorize them ahead of time by assigning a permission set to the app. Without this, Salesforce rejects the JWT assertion with `user hasn't approved this consumer`.</Note>

<Note>Setting **Refresh Token Policy** to **Refresh token is valid until revoked** is also required, even though the JWT Bearer flow never actually issues a refresh token. Salesforce internally checks for a non-expired approval record for the user before issuing an access token, and any other setting causes that approval to age out — which silently breaks the integration later.</Note>

  </Tab>
</Tabs>

> ### 5. Pick a Profile for the integration user

Reviewer: Saw your slack message that either profile or permission set can be associated with an app, we actually already fully document creating a dedicated integration user with a custom permission set. So please create a [reusable snippet](https://www.mintlify.com/docs/create/reusable-snippets) for https://docs.withampersand.com/customer-guides/salesforce#salesforce-integration-users or parts of it, instead of reinventing the wheel here.

Reviewer: Also if you do that, there are already plenty of screenshots in that section.

In Salesforce, a **Profile** determines what a user can access and — importantly for JWT Bearer — which apps they can be pre-authorized for. In step 7 you will add a Profile to the app's **Selected Profiles** list, which pre-authorizes every user on that Profile to act as the JWT `sub` for the app. For that reason, you want the Profile you pick here to be as **narrow as possible** — ideally, one that is used only by the integration user.

<Tabs>
  <Tab title="Create a dedicated integration Profile (recommended)">

Creating a custom Profile scoped to integrations is the cleanest and most secure option. It means only users on this specific Profile can be impersonated by the integration, and you can revoke access instantly by removing the Profile from the app.

1. In **Setup**, search for **Profiles** in the Quick Find box and select **Profiles**.
2. Find a base profile to clone. A good starting point is **Standard User** (or **Minimum Access - Salesforce** for a narrower starting point). Click **Clone** next to it.

> 3. Give the clone a clear name, e.g. `Ampersand Integration Profile`, and click **Save**.

Reviewer: ```suggestion
3. Give the clone a clear name, e.g. `Integration Profile`, and click **Save**.
```

Reviewer: We don't really want the Ampersand name to show up inside of consumer's Salesforce, since it can easily confuse other people at the company who don't know what Ampersand is, leading to this profile being accidentally deleted.

<Note>If you use a shared profile like `System Administrator`, understand the security tradeoff: every admin in your org becomes eligible to be the `sub` of a JWT Bearer exchange for this app. For production deployments, we recommend creating a dedicated integration profile instead.</Note>

  </Tab>
</Tabs>

### 6. Assign the Profile to an integration user

The JWT Bearer flow runs as a specific Salesforce user. Every API call made by the integration is attributed to this user, and its Profile determines what data the integration can read and write, and whether it's allowed to authenticate to the app.

<Tabs>
  <Tab title="Create a new integration user (recommended)">

Creating a dedicated integration user is safer and clearer in audit logs, and it won't break if a teammate leaves the company.

1. In **Setup**, search for **Users** in the Quick Find box and select **Users**.
2. Click **New User**.
3. Fill in the user details:

>    - **First Name** / **Last Name**: e.g. `Ampersand` / `Integration`

Reviewer: ```suggestion
   - **First Name** / **Last Name**: e.g. `Integration` / `User`
```

   - **Email**: use a real mailbox you control. Salesforce will send the welcome email here.

>    - **Username**: must be unique across all Salesforce orgs globally, and must look like an email address (e.g. `ampersand-integration@your-company.com`). It does not need to match any real inbox.

Reviewer: ```suggestion
   - **Username**: must be unique across all Salesforce orgs globally, and must look like an email address (e.g. `integration@your-company.com`). It does not need to match any real inbox.
```

>    - **User License**: pick a license that includes API access. **Salesforce** is the most common choice.

Reviewer: ```suggestion
   - **User License**: pick a license that includes API access, such as **Salesforce** or **Salesforce Integration**. Please note that **Salesforce Platform** is insufficient.
```

   - **Profile**: select the Profile from step 5.
4. Click **Save**.
5. Complete the welcome email flow to set an initial password for the user. JWT Bearer will not use the password, but Salesforce requires the account to be fully activated first.

> <Note>The username you enter here is exactly the value you'll paste into the **Salesforce Username** field when installing the integration. Copy it somewhere you can find it later.</Note>

Reviewer: ```suggestion
<Note>The username you enter here is exactly the value you'll paste into the **Salesforce Username** field when installing the integration. Copy it somewhere that you can find it later.</Note>
```

  </Tab>
  <Tab title="Use an existing user">

If you're reusing an existing Salesforce user, confirm the following before continuing:

1. In **Setup**, open **Users** and find the user.
2. Verify the user is **Active**.
3. Verify the user's **Profile** matches what you picked in step 5. If it doesn't, either switch the user to that Profile, or go back to step 5 and pick the user's actual Profile instead (keeping in mind the security tradeoff of using a shared profile).
4. Note the user's **Username** exactly as shown (including any sandbox suffix like `.sandboxname`) — you will need it later.

> <Note>We still strongly recommend a dedicated integration user over a real human's account. If you're using a human user and they later leave the company, the integration will stop working the moment their user is deactivated.</Note>

Reviewer: ```suggestion
<Note>We recommend a dedicated integration user over a real human's account. If you're using a human user and they later leave the company, the integration will stop working the moment their user is deactivated.</Note>
```

  </Tab>
</Tabs>

### 7. Associate the Profile with the app

> The final setup step is adding the Profile from step 5 to the app's pre-authorized Profiles list. This is what satisfies the **Admin approved users are pre-authorized** OAuth policy from step 4 — it tells Salesforce "any user on this Profile is allowed to use this app headlessly."

Reviewer: ```suggestion
The final setup step is adding the Profile from step 5 to the app's pre-authorized Profiles list. This is what satisfies the **Admin approved users are pre-authorized** OAuth policy from step 4 — it tells Salesforce "any user on this Profile is allowed to use this app."
```

Reviewer: I don't think headlessly is a word lol

<Tabs>
  <Tab title="External Client App">

1. In **Setup**, open **External Client App Manager** and select your External Client App.
2. Go to the **Policies** tab.
3. Expand **App Policies**.
4. Under **Select Profiles**, move the Profile from step 5 from **Available Profiles** to **Selected Profiles** using the arrow button (`>`).
5. Click **Save**.

<Note>The **Select Profiles** picker only appears after you set **Permitted Users** to **Admin approved users are pre-authorized** in step 4. If you don't see it, revisit that step first.</Note>

  </Tab>

>   <Tab title="Connected App (legacy)">

Reviewer: can remove this tab
